### ISSUE TITLE
Character LoRA training produces zero visual effect — LoRA weights near-zero after 3000+ steps

### ISSUE THREAD

**Issue by saved-j** [mrbizarro/phosphene#32]
## Problem

Character LoRA training completes successfully but produces **zero visual effect** at inference. The generated video is identical with and without the LoRA loaded. Tested with multiple training runs, face-cropped images, and various parameters.

## Expected Behavior

Character LoRA should alter the generated face to match the training subject (bald man with beard), similar to how Rosetta and McKayla character LoRAs from CivitAI work at strength 1.0.

## Actual Behavior

All test renders with the character LoRA produce the same output as without the LoRA — a woman with curly hair, completely unrelated to the training subject. LoRA file sizes differ between with/without runs (confirming the LoRA IS loaded), but visual output is identical.

### LoRA Weight Analysis (CRITICAL)

After 3000 steps of training on 50 face-cropped images, the LoRA weights are near-zero:

```
LoRA A per-element std: 0.011 (expected random init: ~0.177)
LoRA B per-element mean: 0.007 (initialized to zeros)
```

The LoRA barely updated from initialization. For comparison, a working character LoRA (Rosetta from CivitAI) produces visible identity transfer at strength 1.0.

### Training Config (v4 — face-cropped attempt)

```yaml
lora:
  alpha: 32
  dropout: 0.0
  rank: 32
  target_modules: [to_q, to_k, to_v, to_out, to_gate_logits, proj_in, proj_out]
model:
  model_path: mlx_models/ltx-2.3-mlx-q4
  training_mode: lora
optimization:
  batch_size: 1
  learning_rate: 0.0001
  optimizer_type: adamw
  scheduler_type: constant
  steps: 3000
```

### Training Dataset

- 50 face-cropped images of the subject (bald man with beard, green eyes)
- Captions auto-generated by Gemma 3 12B in `[VISUAL]: trigger, description` format
- Trigger word: `MV1ct0rs0n`
- Images cropped to 512x512 with face detection (OpenCV Haar cascade)

### Example Caption

```
[VISUAL]: MV1ct0rs0n, The subject gestures emphatically with two fingers raised,
framed in a medium shot against a backdrop of shelves with various devices and
string lights, bathed in warm, slightly dim lighting that emphasizes the depth
of the set.
```

### Test Render Comparison

| Test | LoRA | Strength | Result |
|------|------|----------|--------|
| No LoRA | None | - | Woman with curly hair |
| v3 (full body) | MV1ct0rs0n_v2 | 1.0 | Woman with curly hair (identical) |
| v3 (full body) | MV1ct0rs0n_v2 | 5.0 | Testing |
| v4 (face crop) | MV1ct0rs0n_v2 | 1.0 | Woman with curly hair (identical) |
| Rosetta (CivitAI) | LTX-2.3_sks_rosetta | 1.0 | Correct face |
| McKayla (CivitAI) | McKayla_McMilton_LTX | 1.0 | Correct face |

### Working LoRA Comparison (Rosetta from CivitAI)

Rosetta was trained with `ai-toolkit` (ostris/ai-toolkit) and produces correct identity transfer:

```
Training info: step 2500, epoch 62
Software: ai-toolkit v0.10.2
Resolution: 1024x1024
Tensors: 3264, Modules: 34
Dtype: BF16
```

Our LoRA:
```
Tensors: 1634, Modules: 18
Dtype: F32
Resolution: 512x512
```

Key differences: Rosetta includes audio modules (16 extra module types), uses BF16, trains at 1024x1024.

## What We Tried

1. Verified dev transformer is loaded (symlinked `transformer-dev.safetensors` into Q4 dir, confirmed via logging)
2. Verified LoRA keys match model keys (all 817 lora_A keys match after prefix stripping)
3. Verified LoRA delta survives Q8 quantization (delta/quant_step ratio = 14.4)
4. Fixed TARGET_MODULES (added `to_gate_logits`, `proj_in`, `proj_out`)
5. Fixed venv python symlink (`ln -sf python3.11 python`)
6. Face-cropped all 50 training images (face detection + 80% padding)
7. Tried both short captions (`trigger_simple`: "MV1ct0rs0n, man") and Gemma auto-captions
8. Tried center crop and letterbox crop strategies
9. Verified LoRA is loaded at inference (file sizes differ with/without LoRA)

## Questions

1. What is the expected LoRA weight magnitude after successful training? Our weights are ~0.01 std (vs random init ~0.18).
2. Is there a known issue with `ltx_trainer_mlx` producing near-zero gradients for character LoRAs?
3. Are there additional patches or configurations required beyond what the panel applies?
4. Has anyone successfully trained a character LoRA using the Phosphene panel built-in trainer (not ai-toolkit)?

## Environment

- macOS, M4 Max 64 GB
- Phosphene v3.0.6
- LTX 2.3 Q4 + Q8 dev transformer
- mlx==0.31.1

**Comment by mrbizarro** [comment]
Thanks for the unusually detailed report — the near-zero weights are the smoking gun, and they point at a specific, known cause.

**Root cause: the LoRA trained against the distilled transformer, not the dev one.**

Character LoRA training has to run against the **dev** transformer (`transformer-dev.safetensors`), not the distilled checkpoint. The distilled model uses an 8-step sigma schedule (`DISTILLED_SIGMAS`) that doesn't match the trainer's flow-matching objective, so the gradients that reach the adapters are tiny — which is exactly your `A std 0.011 / B mean 0.007` signature. Same family as Lightricks/LTX-2 #175.

The trap: the Q4 model dir ships the **distilled** file, and the loader's default auto-detect order (`transformer.safetensors` → `transformer-distilled` → `transformer-dev`) picks distilled first — so unless dev is present and explicitly forced, training silently optimizes against the wrong base and produces a dead LoRA.

**This is fixed in current Phosphene (v3.2.5+).** The trainer now forces `transformer-dev.safetensors` and *hard-fails with a download prompt* if it's missing, instead of silently falling back to distilled. Your run completed-but-did-nothing, which is the pre-fix behavior — so you're on an older build.

**To fix:**
1. Update Phosphene (Pinokio → Update; the fix shipped in v3.2.5).
2. Open **Train Character** — the preflight will flag `transformer-dev.safetensors` as missing and let you one-click download it (~11 GB; it isn't shipped by default to save disk for inference-only users).
3. Retrain with the same config. You should see the A-matrix std stay near its init (~0.177) and real identity transfer at strength ~1.0.

If you're already on v3.2.5+ **with** `transformer-dev.safetensors` present and still get near-zero weights, that's a different bug — reopen with your training log (look for the line `training-side override: using transformer-dev.safetensors ...`) and I'll dig in right away.

Closing as resolved-in-current-version, but genuinely ping me if updating + the dev transformer doesn't do it.

**Comment by saved-j** [comment]
> Closing as resolved-in-current-version, but genuinely ping me if updating + the dev transformer doesn't do it.

Hello mrbizarro, thanks, I'll try this right away.
However, while the training is in process, I still wanna ask a bunch of other questions.

It seems that Q4 produces a sharper result as it renders videos in one pass, whereas Q8 takes longer to render, but still produces a softer footage. I figured out it's because Q8 has a 2-pass approach where it first renders in half the resolution, and then upscales it, but this approach fails to outperform Q4 in detail and leaves me with a question: why the hell does it exist then? It's longer, it's softer... What's the point?

But then, the problem is, Phosphene doesn't let users either render videos with LORAs in Q4 or render images with Q8 in one pass to achieve higher footage quality.

ComfyUI with its ... let's say uncomfortable design - however - solves that with a few tricks. 

Now, is there a way to make videos with LORAs on Q4, or force single-pass approach on Q8, or whatever, but just to get a slightly higher quality?

**Comment by saved-j** [comment]
> genuinely ping me if updating + the dev transformer doesn't do it.

now, i've found out that my AI agent (Hermes) messed up with the model files and even though the panel check reported success, the files were not actually okay.

I think the "model check" needs to be a little more conservative and check the model structure as well, and make sure right files are in the right places, and no unnecessary stuff is in the way of success.

Now, I'm finishing the new downloading of the models again (internet in China with VPN sucks), and I'll report the progress later.
